pkg/vm/engine/tae/containers/vector.go

```diff
@@ -123,304 +284,221 @@ func (vec *vector[T]) Equals(o Vector) bool {
 	}
 	return true
 }
-func (vec *vector[T]) IsView() bool                { return vec.impl.IsView() }
-func (vec *vector[T]) Nullable() bool              { return vec.impl.Nullable() }
-func (vec *vector[T]) IsNull(i int) bool           { return vec.impl.IsNull(i) }
-func (vec *vector[T]) HasNull() bool               { return vec.impl.HasNull() }
-func (vec *vector[T]) NullMask() *roaring64.Bitmap { return vec.impl.NullMask() }
-func (vec *vector[T]) Bytes() *Bytes               { return vec.impl.Bytes() }
-func (vec *vector[T]) Data() []byte                { return vec.impl.Data() }
-func (vec *vector[T]) DataWindow(offset, length int) []byte {
-	return vec.impl.DataWindow(offset, length)
+
+func (vec *vector[T]) Foreach(op ItOp, sels *roaring.Bitmap) error {
+	return vec.ForeachWindow(0, vec.Length(), op, sels)
 }
-func (vec *vector[T]) CloneWindow(offset, length int, allocator ...*mpool.MPool) Vector {
-	opts := Options{}
-	if len(allocator) == 0 {
-		opts.Allocator = vec.GetAllocator()
+func (vec *vector[T]) ForeachWindow(offset, length int, op ItOp, sels *roaring.Bitmap) (err error) {
+	err = vec.forEachWindowWithBias(offset, length, op, sels, 0, false)
+	return
+}
+
+func (vec *vector[T]) ForeachShallow(op ItOp, sels *roaring.Bitmap) error {
+	return vec.ForeachWindowShallow(0, vec.Length(), op, sels)
+}
+
+func (vec *vector[T]) ForeachWindowShallow(offset, length int, op ItOp, sels *roaring.Bitmap) error {
+	return vec.forEachWindowWithBias(offset, length, op, sels, 0, true)
+}
+
+func (vec *vector[T]) forEachWindowWithBias(offset, length int, op ItOp, sels *roaring.Bitmap, bias int, shallow bool) (err error) {
```

Comment:
In the previous implementation, many places distinguish whether the current vector contains null and whether the vector in the operand contains null. These are mainly to consider some performance optimization, but now I think they are all removed, and the correctness should be no problem.